coreneuron/io/core2nrn_data_return.cpp

```diff
@@ -145,7 +243,272 @@ void core2nrn_data_return() {
+    (*core2nrn_PreSyn_flag_)(nt.id, presyns_flag_true);
+}
+
+void nrn2core_PreSyn_flag_receive(int tid) {
+    NrnThread& nt = nrn_threads[tid];
+    // turn off all the PreSyn.flag_ as they might have been turned off
+    // on the NEURON side if NEURON integrated a bit.
+    for (int i = 0; i < nt.n_presyn; ++i) {
+        nt.presyns_helper[i].flag_ = 0;  // in case 1 from previous psolve
+    }
+    std::set<int> presyns_flag_true;
+    (*nrn2core_transfer_PreSyn_flag_)(tid, presyns_flag_true);
+    if (presyns_flag_true.empty()) {
+        return;
+    }
+    std::unique_ptr<int[]> pinv_nt;
+    if (nt._permute) {
+        pinv_nt.reset(inverse_permute(nt._permute, nt.end));
+    }
+    for (int i = 0; i < nt.n_presyn; ++i) {
+        PreSyn& ps = nt.presyns[i];
+        PreSynHelper& psh = nt.presyns_helper[i];
+        if (ps.thvar_index_ >= 0) {
+            int index_v = pinv_nt ? pinv_nt[ps.thvar_index_] : ps.thvar_index_;
+            if (presyns_flag_true.erase(index_v)) {
+                psh.flag_ = 1;
+                if (presyns_flag_true.empty()) {
+                    break;
+                }
+            }
+        }
+    }
+}
+
+static void core2nrn_tqueue_item(TQItem* q, NrnThread& nt) {
+    DiscreteEvent* d = (DiscreteEvent*) q->data_;
+    double td = q->t_;
+
+    // potentially several SelfEvent TQItem* associated with same weight index.
+    std::map<int, std::vector<TQItem*>> self_event_weight_map;
+
+    switch (d->type()) {
+        case NetConType: {
+            NetCon* nc = (NetCon*) d;
+            assert(nc >= nt.netcons && (nc < (nt.netcons + nt.n_netcon)));
+            size_t nc_index = nc - nt.netcons;
+            (*core2nrn_NetCon_event_)(nt.id, td, nc_index);
+            break;
+        }
+        case SelfEventType: {
+            SelfEvent* se = (SelfEvent*) d;
+            Point_process* pnt = se->target_;
+            assert(pnt->_tid == nt.id);
+            int tar_type = (int) pnt->_type;
+            int tar_index = pnt->_i_instance;
+            double flag = se->flag_;
+            TQItem** movable = (TQItem**) (se->movable_);
+            int is_movable = (movable && *movable == q) ? 1 : 0;
+            int weight_index = se->weight_index_;
+            // the weight_index is useless on the NEURON side so we need
+            // to convert that to NetCon index  and let the NEURON side
+            // figure out the weight_index. To figure out the netcon_index
+            // construct a {weight_index : [TQItem]} here for any
+            // weight_index >= 0, otherwise send it NEURON now.
+            if (weight_index >= 0) {
+                self_event_weight_map[weight_index].push_back(q);
+            } else {
+                (*core2nrn_SelfEvent_event_noweight_)(
+                    nt.id, td, tar_type, tar_index, flag, is_movable);
+            }
+            break;
         }
+        case PreSynType: {
+            PreSyn* ps = (PreSyn*) d;
+            printf("PreSynType %g\n", td);
```

Comment:
`ps` is not used.
```suggestion
            // nothing to transfer
```

Reply:
Good idea, but printf can pollute logs on MPI runs.

Reply:
LGTM now.